apps/test-app/frontend/src/components/tree-widget/TreeWidget.tsx

```diff
@@ -139,43 +128,136 @@ export function RulesDrivenTreeWidget(props: Props) {
     schemas.addLocater(new ECSchemaRpcLocater(props.imodel.getRpcProps()));
-    const metadataProvider = createMetadataProvider(schemas);
-    const modelsTreeHierarchyProvider = new HierarchyProvider({
-      metadataProvider,
-      hierarchyDefinition: new ModelsTreeDefinition({ metadataProvider }),
-      queryExecutor: createECSqlQueryExecutor(props.imodel),
-    });
+    setQueryExecutor(createECSqlQueryExecutor(props.imodel));
+    setMetadataProvider(createMetadataProvider(schemas));
+  }, [props.imodel]);
+
+  const { value } = useDebouncedAsyncValue(
+    useCallback(async () => {
+      if (metadataProvider && queryExecutor && filter !== "") {
+        setFilteringStatus(FilteringInputStatus.FilteringInProgress);
+        return ModelsTreeDefinition.createInstanceKeyPaths({ metadataProvider, queryExecutor, label: filter });
+      }
+      return [];
+    }, [metadataProvider, queryExecutor, filter]),
+  );
+
+  useEffect(() => {
+    if (metadataProvider && queryExecutor) {
+      const sharedProps = {
+        metadataProvider,
+        hierarchyDefinition: new ModelsTreeDefinition({ metadataProvider }),
+        queryExecutor,
+      };
+      if (value) {
+        setFilteringStatus(FilteringInputStatus.FilteringFinished);
+        setModelsTreeHierarchyProvider(
+          new HierarchyProvider({
+            ...sharedProps,
+            filtering: {
+              paths: value,
+            },
+          }),
+        );
+      } else {
+        setModelsTreeHierarchyProvider(new HierarchyProvider(sharedProps));
+      }
+    }
+  }, [queryExecutor, metadataProvider, value]);
+
+  const dataProvider = useMemo((): TreeDataProvider => {
     return async (node?: TreeNodeItem): Promise<TreeNodeItem[]> => {
       const parent: HierarchyNode | undefined = node ? (node as any).__internal : undefined;
       try {
-        return (await modelsTreeHierarchyProvider.getNodes({ parentNode: parent })).map(parseTreeNodeItem);
+        if (modelsTreeHierarchyProvider) {
+          return (await modelsTreeHierarchyProvider.getNodes({ parentNode: parent })).map(parseTreeNodeItem);
+        }
+        return [];
       } catch (e) {
         // eslint-disable-next-line no-console
         console.error(e);
         return [];
       }
     };
-  }, [props.imodel]);
+  }, [modelsTreeHierarchyProvider]);
   const modelSource = useTreeModelSource(dataProvider);
   const nodeLoader = useTreeNodeLoader(dataProvider, modelSource);
   const eventHandler = useMemo(() => new TreeEventHandler({ nodeLoader, modelSource }), [nodeLoader, modelSource]);
   const treeModel = useTreeModel(modelSource);
+
+  const [heightToUse, setHeightToUse] = useState(0);
+  const treeWidgetHeaderRef = useRef<HTMLDivElement>(null);
+  useEffect(() => {
+    const heightOfHeader = treeWidgetHeaderRef.current?.clientHeight ?? 0;
+    const heightToSet = props.height ? props.height - heightOfHeader : 0;
+    setHeightToUse(heightToSet ?? 0);
+  }, [props.height]);
```

Comment:
Looks like this can be moved into custom hook and reused between both trees:
`const { headerRef, treeHeight } = useTreeHeight(props.heigth);`

Reply:
Added custom hook.